rclpy_components/rclpy_components/component_manager.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2020 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from rclpy.node import Node
+from rclpy.executors import Executor
+from rclpy.logging import get_logger
+from rclpy.exceptions import InvalidNodeNameException, InvalidNamespaceException
+from composition_interfaces.srv import ListNodes, LoadNode, UnloadNode
+try:
+    from importlib.metadata import entry_points
+except ImportError:
+    from importlib_metadata import entry_points
+
+RCLPY_COMPONENTS = 'rclpy_components'
+logger = get_logger('ComponentManager')
+
+
+class ComponentManager(Node):
+
+    def __init__(self, executor: Executor, name="py_component_manager", **kwargs):
+        # TODO Handle the py args equivalent to rclcpp 'NodeOptions'
+        super().__init__(name, **kwargs)
+        self.executor = executor
+        # Implement the 3 services described in
+        # http://design.ros2.org/articles/roslaunch.html#command-line-arguments
+        self.list_node_srv_ = self.create_service(ListNodes, "~/_container/list_nodes", self.on_list_node)
+        self.load_node_srv_ = self.create_service(LoadNode, "~/_container/load_node", self.on_load_node)
+        self.unload_node_srv_ = self.create_service(UnloadNode, "~/_container/unload_node", self.on_unload_node)
+
+        self.components = {}  # key: unique_id, value: full node name and component instance
+        self.unique_id_index = 0
+
+        self.executor.spin()
+
+    def gen_unique_id(self):
+        self.unique_id_index += 1
+        return self.unique_id_index
+
+    def on_list_node(self, req: ListNodes.Request, res: ListNodes.Response):
+        res.unique_ids = [int(key) for key in self.components.keys()]
+        res.full_node_names = [v[0] for v in self.components.values()]
+
+        return res
+
+    def on_load_node(self, req: LoadNode.Request, res: LoadNode.Response):
+        component_entry_points = entry_points().get(RCLPY_COMPONENTS, None)
+        if not component_entry_points:
+            logger.error('No rclpy components registered')
+            res.success = False
+            return res
+
+        component_entry_point = None
+        for ep in component_entry_points:
+            if ep.name == req.plugin_name:
+                component_entry_point = ep
+                break
+
+        if not component_entry_point:
+            logger.error('No rclpy component found by %s' % req.plugin_name)
```

Comment:
Few things here:

1. Instead of logging an error, we can set the `error_string` member of the response object.
2. We should wrap the plugin name in quotes so it's easy to spot empty strings.
3. I think it would be best to be consistent with string formatting throughout the file. I recommend using [f-strings](https://realpython.com/python-f-strings/) since this will likely target Python 3.6+.